File: app/src/main/java/com/xdk78/nimebox/ui/AnimeListFragment.java
package com.xdk78.nimebox.ui;

import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.xdk78.nimebox.R;
import com.xdk78.nimebox.adapter.AnimeListAdapter;
import com.xdk78.nimebox.api.RepositoryProvider;
import com.xdk78.nimebox.model.AnimeList;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import static com.xdk78.nimebox.util.Utils.ANIME_URL;

public class AnimeListFragment extends Fragment {

    private View view;
    private RecyclerView recyclerView;
    private AnimeListAdapter adapter;
    private List<AnimeList> animeList;
    private Unbinder unbinder;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public AnimeListFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        view = inflater.inflate(R.layout.fragment_anime_list, container, false);
        unbinder = ButterKnife.bind(this, view);

        recyclerView = view.findViewById(R.id.anime_list);

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        adapter = new AnimeListAdapter(animeList = new ArrayList<>(), getContext());
        recyclerView.setAdapter(adapter);

        loadAPI();
        return view;
    }

    public void loadAPI() {
        RepositoryProvider repository = new RepositoryProvider();
        compositeDisposable.add(
                repository.provideRepository(ANIME_URL).getAnimeList()
                        .toList()
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeOn(Schedulers.io())
                        .subscribe(
                                animeList -> adapter.addItems(animeList),
                                e -> Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show()
                        )
        );
    }


    @Override
    public void onDestroyView() {
        unbinder.unbind();
        compositeDisposable.clear();
        super.onDestroyView();
    }
}

File: app/src/main/java/com/xdk78/nimebox/ui/AnimeListFragment.kt
package com.xdk78.nimebox.ui

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.xdk78.nimebox.R
import com.xdk78.nimebox.adapter.AnimeListAdapter
import com.xdk78.nimebox.base.BaseFragment
import com.xdk78.nimebox.mvp.model.AnimeListModel
import com.xdk78.nimebox.mvp.presenter.AnimeListPresenter
import com.xdk78.nimebox.mvp.view.AnimeListView
import kotlinx.android.synthetic.main.fragment_anime_list.*
import org.jetbrains.anko.support.v4.find
import java.io.IOException
import javax.inject.Inject

class AnimeListFragment : BaseFragment(), AnimeListView {

    lateinit var adapter: AnimeListAdapter

    @Inject
    lateinit var presenter: AnimeListPresenter

    lateinit var recyclerView: RecyclerView

    companion object {
        fun newInstance(): AnimeListFragment {
            return AnimeListFragment()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_anime_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        swiperefresh.setOnRefreshListener {
            presenter.subscribe(this)
            swiperefresh.isRefreshing = true
            presenter.loadData()
        }
        presenter.subscribe(this)
        presenter.loadData()
    }


    override fun onItemLoaded(items: List<AnimeListModel>) {
        recyclerView = find(R.id.anime_list)
        recyclerView.layoutManager = LinearLayoutManager(context)
        adapter = AnimeListAdapter(context!!, items)
        recyclerView.adapter = adapter

        swiperefresh.isRefreshing = false
    }

    override fun onError(e: Throwable?) {
        when (e) {
            is IOException -> {
                Log.e("api", e.message)
                Snackbar.make(this.view!!, e.toString(), Snackbar.LENGTH_LONG).show()
                swiperefresh.isRefreshing = false
            }
            else -> Log.e("api", "List is empty")
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.unsubscribe()
    }

    override fun onDetach() {
        super.onDetach()
        presenter.unsubscribe()
    }
}